src/index.ts

```diff
@@ -854,6 +870,64 @@ export abstract class PinejsClientCore<PinejsClient> {
 		}
 	}
 
+	private canRetryDefaultHandler(err: any) {
+		const code = err?.statusCode;
+		return !code || code === 429 || (code >= 500 && code < 600);
+	}
+
+	protected async callWithRetry<T>(
+		fnCall: () => Promise<T>,
+		retry: RetryParameters,
+	): Promise<T> {
+		if (!retry) {
+			return await fnCall();
+		}
+
+		const defaultRetry = this.retry || {};
+		const minDelayMs = retry.minDelayMs ?? defaultRetry.minDelayMs ?? 0;
+		const maxDelayMs = retry.maxDelayMs ?? defaultRetry.maxDelayMs ?? 0;
+		const maxAttempts = retry.maxAttempts ?? defaultRetry.maxAttempts ?? 0;
+
+		if (
+			minDelayMs <= 0 ||
+			maxDelayMs <= 0 ||
+			maxAttempts <= 0 ||
+			minDelayMs > maxDelayMs
+		) {
+			console.warn('pinejs-client invalid retry parameters');
+			return await fnCall();
+		}
+
+		const onRetryHandler = retry.onRetry ?? defaultRetry.onRetry;
+
+		let attempt = 0;
+		let prevErr: any;
+
+		const canRetryHandler =
+			retry.canRetry ?? defaultRetry.canRetry ?? this.canRetryDefaultHandler;
+
+		while (attempt < maxAttempts) {
+			try {
+				if (attempt > 0) {
+					const delayMs = Math.min(2 ** attempt * minDelayMs, maxDelayMs);
+					onRetryHandler?.(prevErr, delayMs, attempt + 1, maxAttempts);
+					await new Promise<void>((resolve) => {
+						setTimeout(resolve, delayMs);
+					});
+				}
+				return await fnCall();
+			} catch (err) {
+				if (!canRetryHandler(err)) {
+					throw err;
+				}
+				prevErr = err;
+			}
+			attempt++;
+		}
```

Comment:
```suggestion
		while (attempt < maxAttempts) {
			try {
				return await fnCall();
			} catch (err) {
				if (!canRetryHandler(err)) {
					throw err;
				}
				const delayMs = Math.min(2 ** attempt * minDelayMs, maxDelayMs);
				onRetryHandler?.(err, delayMs, attempt + 1, maxAttempts);
				await new Promise<void>((resolve) => {
					setTimeout(resolve, delayMs);
				});
				prevErr = err;
			}
			attempt++;
		}
```

Reply:
Just checking if you really want to do this?  This will cause an additional delay in the case of all attempts running out and having to throw the error at the end.  It will also mean that we log a retry for that final case when there is no retry to follow.  This feels wrong to me.